cloudmarker/stores/mongodbstore.py

```diff
@@ -0,0 +1,100 @@
+"""MongoDB store plugin."""
+
+
+import logging
+
+from pymongo import MongoClient
+
+_log = logging.getLogger(__name__)
+
+
+class MongoDBStore:
+    """A plugin to store records on the MongoDB."""
+
+    def __init__(self, host, port, username=None, password=None, db='gcp',
+                 buffer_size=1000):
+        """Initialize object of this class with specified parameters.
+
+        Arguments:
+            host (str): MongoDB instance host.
+            port (int): MongoDB instance port.
+            username (str): MongoDB instance auth username.
+            password (str): MongoDB instance auth password.
+            db (str): MongoDB database name, default: ``gcp``
+            buffer_size (int): Max number of records to hold in memeory for
+                each record_type.
+
+        """
+        # pylint: disable=too-many-instance-attributes
+
+        self._mongodb_host = host
+        self._mongodb_port = port
+        self._mongodb_username = username
+        self._mongodb_password = password
+        self._db = db
+
+        self._buffer_size = buffer_size
+        self._buffer = {}
+
+        if not (username and password):
```

Comment:
`username` and `password` should be a required param as database will fail early instead of failing when the first `_flush()` will be called. 
```
>>> from pymongo import MongoClient
>>> client = MongoClient(host="0.0.0.0", connect=False)
>>> db = client["mongo-test"]
>>> db.list_collection_names()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/p0n002h/.venv/py-orm/lib/python3.7/site-packages/pymongo/database.py", line 690, in list_collection_names
    nameOnly=True)]
  File "/Users/p0n002h/.venv/py-orm/lib/python3.7/site-packages/pymongo/database.py", line 677, in list_collections
    **kwargs)
  File "/Users/p0n002h/.venv/py-orm/lib/python3.7/site-packages/pymongo/database.py", line 631, in _list_collections
    session=tmp_session)["cursor"]
  File "/Users/p0n002h/.venv/py-orm/lib/python3.7/site-packages/pymongo/database.py", line 514, in _command
    client=self.__client)
  File "/Users/p0n002h/.venv/py-orm/lib/python3.7/site-packages/pymongo/pool.py", line 579, in command
    unacknowledged=unacknowledged)
  File "/Users/p0n002h/.venv/py-orm/lib/python3.7/site-packages/pymongo/network.py", line 150, in command
    parse_write_concern_error=parse_write_concern_error)
  File "/Users/p0n002h/.venv/py-orm/lib/python3.7/site-packages/pymongo/helpers.py", line 155, in _check_command_response
    raise OperationFailure(msg % errmsg, code, response)
pymongo.errors.OperationFailure: command listCollections requires authentication
```

Reply:
Connection to MongoDB server may or may not be authenticated. If the default keyword argument is None, then the connect call to mongo will fail with the following error.

```
  File "/Users/r0s00s4/.venv/cloudmarker/lib/python3.7/site-packages/pymongo/client_options.py", line 38, in _parse_credentials
    mechanism, source, username, password, options, database)
  File "/Users/r0s00s4/.venv/cloudmarker/lib/python3.7/site-packages/pymongo/auth.py", line 107, in _build_credentials_tuple
    raise ConfigurationError("%s requires a username." % (mech,))
pymongo.errors.ConfigurationError: SCRAM-SHA-256 requires a username.
``` 
The last line of trace shows the reason for failure to connect the call. In background the PyMongo will create a URI like this `"mongodb://user:password@example.com/?authSource=the_database&authMechanism=SCRAM-SHA-256"` if None is passed, the call would look something like `"mongodb://:@example.com/?authSource=the_database&authMechanism=SCRAM-SHA-256"` which is a malformed URI to connect to.

The alternative to this is to keep the empty string in username and password in the keyword argument.

Reply:
@prateeknischal What's the behaviour with empty string value for `username` and `password` when you connect with MongoClient. Does the MongoDB support empty username and password? If it does, then we would have to investigate why the Python API does not support it.

I am yet to experiment with MongoDB, so right now, I am just going to rely on your knowledge to form my own opinion about your other comment related to `MongoDBStore.__init__` parameter order.

Reply:
The python API does support not having to initialise `username` and `password` as shown in this function definition http://api.mongodb.com/python/current/api/pymongo/database.html#pymongo.database.Database.authenticate. @prateeknischal and I tested it out when I happened to come across this function. Even though it is deprecated, it still means that Mongo DB allows storing without authentication.

Reply:
@susam If you set up the mongo DB server without any username or password, then python API accepts both `None` and `""` as both `username` and `password`.

Reply:
To test that please update [docker-compose.yml](https://gecgithub01.walmart.com/p0n002h/py-orm/blob/master/mongo/docker-compose.yml) file by removing the `enviroment` sections in both the containers, update `username` and `password` in [mongo_test.py](https://gecgithub01.walmart.com/p0n002h/py-orm/blob/master/mongo/mongo_test.py) file and then run:
```
. venv
make mongo
```

Reply:
@r0hi7, just got curious on the URL format. 
`user` and `password` are optional params in the [RFC 1738](https://www.ietf.org/rfc/rfc1738.txt), so `mongodb://:@example.com/?authSource=the_database&authMechanism=SCRAM-SHA-256` is still valid.
```
user
    An optional user name. Some schemes (e.g., ftp) allow the
    specification of a user name.

password
    An optional password. If present, it follows the user
    name separated from it by a colon.
```